File: BinaryTreeRightSideView.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
# Time Complexity - O(n)
# Space Complexity - O(n)
class Solution(object):
    def rightSideView(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        if not root:
            return []
        rightsideView = []
        queue = [root]
        temp = 0
        while queue:
            l = len(queue)
            for _ in range(l):
                curr = queue.pop(0)
                temp = curr.val
                if curr.left:
                    queue.append(curr.left)
                if curr.right:
                    queue.append(curr.right)
            rightsideView.append(temp)
        return rightsideView

File:  BinaryTreeRightSideView.swift
//After class
// Time Complexity : (O)N
// Space Complexity : (O)h
// Did this code successfully run on Leetcode : yes
// Use recursive postorder traversal appraoch to add right view by checking the depth with result array. Cince we are traversing on the right side we always first visit the right view even it has only left node.

class Solution {
    var result: [Int] = [Int]()
    func rightSideView(_ root: TreeNode?) -> [Int] {
        postorderTraversal(root, 0)
        return result
    }
    
    private func postorderTraversal(_ root: TreeNode?, _ depth: Int) {
        //base
        if root == nil {
            return
        }
        if depth == result.count {
            result.append(root!.val)
        }
        //logic
        postorderTraversal(root?.right, depth+1)
        postorderTraversal(root?.left, depth+1)
    }
}